### PR TITLE
schema/config-linux: add min/max limit for resource memory and cpu

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com

### CODE REVIEW

File: schema/config-linux.json [hunk 1]
```diff
@@ -170,11 +170,14 @@
                             },
                             "period": {
                                 "id": "https://opencontainers.org/schema/bundle/linux/resources/cpu/period",
-                                "$ref": "defs.json#/definitions/uint64Pointer"
+                                "type": "integer",
+                                "minimum": 1000,
+                                "maximum": 1000000
```
Comment (cyphar) [inline schema/config-linux.json]: Where do these magic numbers come from? If they come from the kernel, please document them in the `config-linux.md` file.

File: schema/config-linux.json [hunk 2]
```diff
@@ -239,7 +242,9 @@
                             },
                             "swappiness": {
                                 "id": "https://opencontainers.org/schema/bundle/linux/resources/memory/swappiness",
-                                "$ref": "defs.json#/definitions/uint64Pointer"
+                                "type": "integer",
+                                "minimum": 0,
+                                "maximum": 100
```
Comment (cyphar) [inline schema/config-linux.json]: The minimum is `-1` for `swappiness`.

File: schema/config-linux.json [hunk 3]
```diff
@@ -170,11 +170,14 @@
                             },
                             "period": {
                                 "id": "https://opencontainers.org/schema/bundle/linux/resources/cpu/period",
-                                "$ref": "defs.json#/definitions/uint64Pointer"
```
Comment (cyphar) [inline schema/config-linux.json]: These absolutely need to be `uint64Pointer`s. The reason for this is explained (in great detail) in #233.

File: schema/config-linux.json (hunk 1 above)
Comment (dqminh) [inline schema/config-linux.json]: https://github.com/torvalds/linux/blob/01ea443982203fcdee1250ab630ab6516f0a16e6/kernel/sched/core.c#L8326-L8327

File: schema/config-linux.json (hunk 2 above)
Comment (wking) [inline schema/config-linux.json]: On Thu, Aug 18, 2016 at 02:41:21AM -0700, Aleksa Sarai wrote:
> > @@ -239,7 +242,9 @@
> >                              },
> >                              "swappiness": {
> >                                  "id": "https://opencontainers.org/schema/bundle/linux/resources/memory/swappiness",
> > -                                "$ref": "defs.json#/definitions/uint64Pointer"
> > +                                "type": "integer",
> > +                                "minimum": 0,
> > +                                "maximum": 100
> 
> The minimum is `-1` for `swappiness`.

The Go type is unsigned, so we're blocking -1 there.  More discussion
on this issue, whether we need to allow -1, and potential alternatives
in #499.

File: schema/config-linux.json (hunk 3 above)
Comment (Mashimiao) [inline schema/config-linux.json]: Sorry, I read #233 in detail. But I think `uint64Pointer` can't satisfy the special requirements for swappiness, period and quota.
I updated PR to support `null` type for nil pointer.

**Comment by vbatts** [comment]
@cyphar thoughts on this now?

**Comment by tianon** [comment]
friendly ping @cyphar :heart: :pray:

**Comment by cyphar** [comment]
I'm confused what this will look like as an actual Go structure.

**Comment by wking** [comment]
On Fri, Nov 04, 2016 at 08:01:49PM -0700, Aleksa Sarai wrote:
> I'm confused what this will look like as an actual Go structure.

They'll be *uint64, just like they are now.  The current uint64Pointer
definition is just like what @Mashimiao is adding here [1], except
with different limits on the integer case [2].

[1]: https://github.com/opencontainers/runtime-spec/blob/41ef05e714f2e629796664aae640c777b2d34e9a/schema/defs.json#L69-L78
[2]: https://github.com/opencontainers/runtime-spec/blob/41ef05e714f2e629796664aae640c777b2d34e9a/schema/defs.json#L39-L43

**Comment by crosbymichael** [comment]
How is this even used at the system level?

**Comment by wking** [comment]
On Mon, Nov 07, 2016 at 09:16AM -0800,, Michael Crosby wrote:
> How is this going to used at the system level? 

Is that “who consumes the JSON Schema”?  There's currently `schema/validate.go` in this repo which lets anyone check their JSON against the official schema.  And I've floated opencontainers/runtime-tools#197 to lean on those JSON Schemas for runtime-tools' validation.  I'd expect folks building a new config or checking a config they've just received from someone else will want to validate that config before attempting to launch a container from it.

**Comment by Mashimiao** [comment]
@opencontainers/runtime-spec-maintainers  PTAL

**Comment by crosbymichael** [comment]
This would be validating system settings.  My feeling is that we don't need to spec out the kernel, its not our job.

REJECT 

from me

**Comment by wking** [comment]
> This would be validating system settings. My feeling is that we don't need to spec out the kernel, its not our job.

I don't think we're ready for a hard line on that issue, since we currently have a fair amount of in-spec kernel documentation (e.g. [here][1], [here][2], [here][3], …).  But maybe @jlbutler will get us pointed more clearly in his reroll of #673.

And if we want to punt docs to the kernel, we should probably require that `linux.resources.cpu.period` have the same semantics as [`cpu.cfs_period_us`][4].  Otherwise it is not clear what we're punting to.

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0-rc3/config-linux.md#L87
[2]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0-rc3/config-linux.md#L115
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0-rc3/config-linux.md#L216
[4]: https://www.kernel.org/doc/Documentation/scheduler/sched-bwc.txt

**Comment by Mashimiao** [comment]
@crosbymichael  we also validate range of blkio.weight in schema

**Comment by mrunalp** [comment]
Agree with @crosbymichael, Let's not add limits for values.

**Pull request closed without merging**